test/integration_test_wrapper.py

```diff
@@ -0,0 +1,117 @@
+parser.add_argument('--mkosi-image-name', required=True)
+parser.add_argument('--mkosi-output-path', required=True, type=Path)
+parser.add_argument('--test-number', required=True)
+parser.add_argument('--setup-selinux', default=False, action='store_true')
+parser.add_argument('--skip-shutdown', default=False, action='store_true')
+parser.add_argument('mkosi_args', nargs="*")
+
+def main():
+    logging_basic_config(level=LOG_DEBUG)
+    args = parser.parse_args()
+
+    test_unit_name = f"testsuite-{args.test_number}.service"
+    # Machine names shouldn't have / since it's used as a file name
+    # and it must be a valid hostname so 64 chars max
+    machine_name = args.test_name.replace('/', '_')[:64]
+
+    log.debug(f"test name: {args.test_name}\n"
+              f"test number: {args.test_number}\n"
+              f"image: {args.mkosi_image_name}\n"
+              f"mkosi output path: {args.mkosi_output_path}\n"
+              f"mkosi args: {args.mkosi_args}\n"
+              f"skip shutdown: {args.skip_shutdown}")
+
+    with ExitStack() as stack:
+
+        journal_file = Path(f"{machine_name}.journal").absolute()
+        log.debug(f"Capturing journal to {journal_file}")
+
+        mkosi_args = [
+            'mkosi',
+            '--directory', Path('..').resolve(),
+            '--output-dir', args.mkosi_output_path.absolute(),
+            '--machine', machine_name,
+            '--image', args.mkosi_image_name,
+            '--format=disk',
+            '--ephemeral',
+            '--forward-journal', journal_file,
+            '--kernel-command-line-extra',
+            ' '.join([
+                *BUILTIN_KERNEL_COMMAND_LINE_ARGS,
+                *(
+                    ['apparmor=0', 'selinux=1', 'enforcing=0', 'lsm=selinux']
+                    if args.setup_selinux
+                    else ['apparmor=0', 'selinux=0', 'enforcing=0']
+                ),
+                f"SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-{args.test_number}.units:/usr/lib/systemd/tests/testdata/units:",
+                'systemd.unit=testsuite.target',
+                f"systemd.wants={test_unit_name}",
+            ]),
+            *(
+                [f"--credential=systemd.unit-dropin.{test_unit_name}={quote(TEST_EXIT_DROPIN)}"]
+                if not args.skip_shutdown
+                else []
+            ),
+            *args.mkosi_args,
+        ]
+
+        mkosi_args += ['qemu']
+
+        if log.isEnabledFor(LOG_DEBUG):
+            log.debug(f"Running {' '.join(quote(str(a)) for a in mkosi_args)}")
+
+        try:
+            run(mkosi_args, check=True)
+        except SystemExit as e:
+            if e.code not in (0, 77):
+                log.debug("Attempting journalctl to discover test failure")
+                run(['journalctl',
+                     '--boot',
+                     '--unit', test_unit_name,
+                     '--file', journal_file.name,
+                    ], stdin=DEVNULL)
```

Comment:
We should also log the location of the journal so that users can invoked `journalctl` with different options if they would like.

But I think this might make more sense as part of `mkosi` itself. I would replace this debugging logic for now and just make sure we log the path of the journal as well as this command so users just have to copy paste it to run it themselves.

This also allows us to replace usage of `subprocess.run()` with `os.execve()` which will simplify the script as we don't need to do any child process handholding.

Reply:
I changed this to list the filters used by the bash test suite since it's a lot more useful for finding errors than just `--unit`.